### ISSUE TITLE
wrong number of columns in IFMatrix

### ISSUE THREAD

**Issue by jseabold** [dmlc/xgboost#358]
I'm a bit at a loss here. I can replicate this on linux with gcc 4.8.2 but not 4.6.3. NumPy version is 1.9.2 but shouldn't matter?

```
import numpy as np
from xgboost import XGBClassifier

X = np.random.randn(100, 10)
y = np.random.randint(2, size=100)

mod = XGBClassifier(colsample_bytree=.1)
mod.fit(X, y)
```

This fails on 777c5ce992187bf51920029f6a73012afcb17595 with `colsample_bytree=0.1 is too small that no feature can be included` but it should not. Since `num_col` should be 10, it should select one column, but `num_col` is 9 at this point in the code for some reason. I don't see anything yet as to why.

I'm compiling `cxx11=1`, `-O0`, and `-g` to debug, but it happens with the default too.

**Comment by tqchen** [comment]
This was due to the default 0 missing value. The highest columns gets all 0 I guess, and the construction ignores the highest column(due to all missing). Maybe we should change the construction from dense array to defaults to nan as missing

**Comment by jseabold** [comment]
I'm not sure I follow. No column is all 0 in X or y and the default missing for the scikit-learn API should be np.nan. When I was stepping through the code, I saw that `nan_missing` was true . Also seemed odd that I had trouble replicating on different machines with as far as I know the compiler version being the only thing that's different.

**Comment by tqchen** [comment]
what you mean is when you print out num_col at the point of error, it was 9 instead of 10? In such case, we might have to trace the IFMatrix construction process to see if there is anything suspicious happening

**Comment by jseabold** [comment]
Yeah, that's what I mean. It looks like it has the right number of
columns at the end of XGDMatrixCreateFromMat. I thought maybe
something was going on when the label is appended, but nothing looks
odd to me in the code. But when it checks the num_col for the
`colsample_bytree` check it's no longer 10. So .1 * 9 fails the check.

**Comment by tqchen** [comment]
Can you do a bit investigations? Say trace the DMatrix.info.num_col and IFMatrix.NumCol() at different time point to see where this changes

**Comment by jseabold** [comment]
I spent a while doing this with gdb today and reading the code, and I
didn't see anything obvious. I'll need to do a more careful investigation
but don't know when I'll have time.

Are you unable to replicate? I couldn't trigger it with my gcc build on OSX
(not sure version right now). Only in a docker container and in a
virtualbox using ubuntu with the mentioned version of gcc.

**Comment by tqchen** [comment]
OK, it turns out to be a bug that i overlooked, i fixed it in most recent push.

**Comment by tqchen** [comment]
This do calls for some need of CI and regression test. Currently we setup a testcases for R in the travis, are you interested in setting up a python version possibly? we can reuse the demo scripts

**Comment by jseabold** [comment]
Cross-link the commit. 561e51871ee2e50a4f8361799ed0fae2401b5c0d

I'll see about getting travis setup and adding at least some regression tests. Writing regression tests is how I'm discovering these things.

**Comment by tqchen** [comment]
great!

**Comment by jseabold** [comment]
BTW, I've found it kind of cumbersome to write meaningful regression tests in light of #310.

**Comment by tqchen** [comment]
It depends on whether we want exact equal assert to static data or we want to test whether accuracy on certain data reach some level. The latter could be more meaningful to ML programs

**Comment by jseabold** [comment]
I am talking about failures using correct floating point testing. The
failures I have regularly been seeing happen even using very loose
tolerance levels. Perhaps we can revisit when I get travis running and can
demonstrate this. I'd be interested to get your opinion.
